[Fixes #21927] Compress (Un)BoundedSourceAsSdfWrapper element and restriction coders

lukecwik opened the pull request:
A typical BoundedSource may be split into many BoundedSource instances during initial splitting. Doing a simple test of the BigtableSource shows that encoding 10 instances after splitting took on average 102660 bytes while compressing each instance separately after encoding took 1639 bytes for a >60x improvement.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

lukecwik commented:
It would be great if we could do this for unbounded source as well but that would break pipeline update compatibility.

lukecwik commented:
R: @kileys

lukecwik commented:
CC: @pabloem Would be great to get into the 2.40 release as a patch.

pabloem commented:
Run Java PreCommit

pabloem commented:
Run Java PostCommit

pabloem commented:
I've made a change for this path in Python before (I used a memoizing coder for encoding the source, because the pickling coder was extremely slow, and it was pickling the same BoundedSource object over and over).

Has this been reported by others? Is this causing trouble for someone?

lukecwik commented:
> I've made a change for this path in Python before (I used a memoizing coder for encoding the source, because the pickling coder was extremely slow, and it was pickling the same BoundedSource object over and over).
> 
> Has this been reported by others? Is this causing trouble for someone?

Yes, came up as a customer issue when using BigtableIO.

lukecwik commented:
Note that I updated and added the streaming fix as well. It turns out that changing bounded was going to break pipeline update compatibility for portable pipeline users but we should get this in before we make Java on Dataflow Prime GA.

pabloem commented:
LGTM

lukecwik commented:
Run Java_Examples_Dataflow_Java11 PreCommit

lukecwik commented:
Run Java PreCommit

pabloem commented:
Run Java PreCommit
